sql/src/main/java/org/apache/druid/sql/calcite/run/NativeQueryMaker.java

```diff
@@ -164,6 +169,70 @@ private List<Interval> findBaseDataSourceIntervals(Query<?> query)
                 .orElseGet(query::getIntervals);
   }
 
+  private boolean queryHasTimeFilter(Query<?> query)
+  {
+    DataSource dataSource = query.getDataSource();
+    if (dataSource instanceof InlineDataSource) {
+      return true;
+    }
+    if (dataSource instanceof JoinDataSource) {
+      return joinDataSourceQueryHasTimeFilter(query);
+    }
+    return isIntervalNonEternity(findBaseDataSourceIntervals(query));
+  }
+
+  /**
+   * Checks if a join query has a valid time filter by inspecting parts of the joins and any subqueries used within.
+   * If the left datasource is a Table datasource, we require a timefilter on the top level query and a time filter on
+   * the right datasource, else we check for time filter on the left and right datasources making up the join recursively
+   */
+  private boolean joinDataSourceQueryHasTimeFilter(Query<?> query)
+  {
+    Preconditions.checkArgument(query.getDataSource() instanceof JoinDataSource);
+    JoinDataSource joinDataSource = (JoinDataSource) query.getDataSource();
+    if (joinDataSource.getLeft() instanceof TableDataSource) {
+      // Make sure we have a time filter on the base query since we have a concrete TableDataSource on the left
+      // And then make sure that right also has a time filter
+      if (isIntervalNonEternity(query.getIntervals())) {
+        return dataSourceHasTimeFilter(joinDataSource.getRight());
+      } else {
+        return false;
+      }
+    }
+    // Top level query does not have a time filter, check if all subqueries have a time filter
+    return joinDataSourceHasTimeFilter(joinDataSource);
+  }
+
+  private boolean joinDataSourceHasTimeFilter(JoinDataSource joinDataSource)
+  {
+    return dataSourceHasTimeFilter(joinDataSource.getLeft()) && dataSourceHasTimeFilter(joinDataSource.getRight());
+  }
+
+  private boolean dataSourceHasTimeFilter(DataSource dataSource)
+  {
+    if (dataSource instanceof InlineDataSource) {
+      return true;
+    }
+    if (dataSource instanceof QueryDataSource) {
+      return queryHasTimeFilter(((QueryDataSource) dataSource).getQuery());
+    }
+    if (dataSource instanceof JoinDataSource) {
+      return joinDataSourceHasTimeFilter((JoinDataSource) dataSource);
+    }
+    if (dataSource.getAnalysis().getBaseQuerySegmentSpec().isPresent()) {
+      return isIntervalNonEternity(dataSource.getAnalysis()
+                                             .getBaseQuerySegmentSpec()
+                                             .map(QuerySegmentSpec::getIntervals)
+                                             .get());
+    }
+    return false;
```

Comment:
the pattern of this method makes me wonder if it would be better to have a method on `DataSource` instead (with a `default` implementation to return false) so that every relevant subclass could implement it correctly?

Reply:
Thanks for takin a look @kgyrtkirk . Makes sense to have this at the DataSource level. Made the necessary changes.